internal/command/program_resolver.go

```diff
@@ -0,0 +1,319 @@
+package command
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"slices"
+	"strings"
+
+	"mvdan.cc/sh/v3/syntax"
+)
+
+type ProgramResolverMode uint8
+
+const (
+	// unspecified is auto (default) which prompts for all unresolved
+	// subsequent runs will likely resolve via session
+	ProgramResolverModeAuto ProgramResolverMode = iota
+	// always prompt even if resolved
+	ProgramResolverModePrompt
+	// don't prompt whatsover even unresolved is resolved
+	ProgramResolverModeSkip
+)
+
+type ProgramResolverSource func() []string
+
+func ProgramResolverSourceFunc(env []string) ProgramResolverSource {
+	return func() []string {
+		return env
+	}
+}
+
+// ProgramResolver uses a list of ProgramResolverSource to resolve environment variables
+// found in a shell program. The result contains all found environment variables.
+// If the env is in any source, it is considered resolved. Otherwise, it is marked
+// as unresolved.
+type ProgramResolver struct {
```

Comment:
Overall, I think the implementation would be easier to reason about if `ProgramResolver` didn't contain `mode`. Instead, there would be `ProgramResolverWithMode` which would look like this:

```go
struct ProgramResolverWithMode {
  *ProgramResolver
  Mode ProgramResolverMode
}
```